File: Tree/Binary Tree/binary tree.py
class Node:
    def __init__(self, value):
        self.data = value
        self.right = None
        self.left = None


class binaryTree:
    def __init__(self):
        self.root = None

    def insertValue(self, value):
        newNode = Node(value)
        if self.root is None:
            self.root = newNode
            return
        tree = []
        tree.append(self.root)
        while len(tree):
            temp = tree[0]
            tree.pop(0)
            if (temp.left is None):
                temp.left = newNode
                break
            else:
                tree.append(temp.left)
            if (temp.right is None):
                temp.right = newNode
                break
            else:
                tree.append(temp.right)

    def printLevelOrder(self):
        tree = []
        tree.append(self.root)
        while len(tree):
            temp = tree[0]
            if temp is None:
                return
            print(temp.data)
            tree.pop(0)
            tree.append(temp.left)
            tree.append(temp.right)

    def LeftView(self):
        tree = []
        tree.append(self.root)
        while len(tree):
            temp = tree[0]
            tree.pop(0)
            if temp is None:
                return
            print(temp.data, end="")
            tree.append(temp.left)


if __name__ == "__main__":
    tree = binaryTree()
    # arr = list(map(int, input().split()))
    arr = [1, 2, 3, 4, 5]
    for i in arr:
        tree.insertValue(i)
    tree.printLevelOrder()
    tree.LeftView()

File: Tree/Binary Tree/binary_tree.cpp
#include <iostream>
#include <queue>
#include <stack>
using namespace std;

class node
{
public:
    int data;
    node *left;
    node *right;

    node(int d)
    {
        this->data = d;
        this->left = NULL;
        this->right = NULL;
    }
};

node *buildTree(node *root)
{
    cout << "Enter the data for Node" << endl;
    int data;
    cin >> data;
    root = new node(data);
    if (data == -1)
        return NULL;

    cout << "Enter data left of " << data << endl;
    root->left = buildTree(root->left);

    cout << "Enter data right of " << data << endl;
    root->right = buildTree(root->right);

    return root;
}

void levelOrderTraversal(node *root)
{
    queue<node *> q;
    q.push(root);
    q.push(NULL);
    while (!q.empty())
    {
        node *temp = q.front();
        q.pop();
        if (temp == NULL)
        {
            // new level started
            cout << endl;
            if (!q.empty()) // deque is not empty
                q.push(NULL);
        }
        else
        {
            cout << temp->data << " ";
            if (temp->left)
                q.push(temp->left);
            if (temp->right)
                q.push(temp->right);
        }
    }
}

void reverseLevelOrderTraversal(node *root)
{
    queue<node *> q;
    stack<int> s;
    q.push(root);
    q.push(NULL);
    while (!q.empty())
    {
        node *temp = q.front();
        q.pop();
        if (temp == NULL)
        {
            // new level started
            if (!q.empty())
            {
                // deque is not empty
                q.push(NULL);
                s.push(-1);
            }
        }
        else
        {
            s.push(temp->data);
            if (temp->right)
                q.push(temp->right);
            if (temp->left)
                q.push(temp->left);
        }
    }
    int temp;
    while (!s.empty())
    {
        temp = s.top();
        s.pop();
        if (temp != -1)
            cout << temp << " ";
        else
            cout << endl;
    }
}

int maxDepth(node *root)
{
    if (root == NULL)
        return 0;

    int lh = maxDepth(root->left);
    int rh = maxDepth(root->right);

    return 1 + max(lh, rh);
}

void inOrder(node *root)
{
    if (root == NULL)
        return;
    inOrder(root->left);
    cout << root->data << " ";
    inOrder(root->right);
}

void preOrder(node *root)
{
    if (root == NULL)
        return;
    cout << root->data << " ";
    inOrder(root->left);
    inOrder(root->right);
}

void postOrder(node *root)
{
    if (root == NULL)
        return;
    inOrder(root->left);
    inOrder(root->right);
    cout << root->data << " ";
}

int isBalanced(node *root)
{
    if (root == NULL)
        return 0;
    int lh = isBalanced(root->left);
    if (lh == -1)
        return -1;
    int rh = isBalanced(root->right);
    if (rh == -1)
        return -1;
    cout << endl
         << lh << " " << rh << " " << root->data << " " << abs(lh - rh);
    if (abs(lh - rh) > 1)
        return -1;
    return 1 + max(lh, rh);
}

int main()
{
    node *root = NULL;
    // create a binary tree
    root = buildTree(root);

    // 1 3 7 -1 -1 11 -1 -1 5 17 -1 -1 -1

    // level order
    cout << "levelOrderTraversal" << endl;
    levelOrderTraversal(root);
    cout << "reverseLevelOrderTraversal" << endl;
    reverseLevelOrderTraversal(root);

    cout << endl
         << "inOrder -> ";
    inOrder(root);

    cout
        << endl
        << "preOrder -> ";
    preOrder(root);

    cout << endl
         << "postOrder -> ";
    postOrder(root);

    int h = maxDepth(root);
    cout << endl
         << "Height" << h;
    // int v = isBalanced(root);
    int height = isBalanced(root);
    cout << endl
         << "isBalance " << height << endl;
    return 0;
}
// 1 2 3 4 -1 -1 4 -1 -1 3 -1 -1 2 -1 -1
// 3 9 -1 -1 20 15 -1 -1 7 -1 -1